packages/vm/src/memory.rs

```diff
@@ -39,12 +40,11 @@ pub fn read_region(memory: &wasmer::Memory, ptr: u32, max_length: usize) -> VmRe
 
     match WasmPtr::<u8, Array>::new(region.offset).deref(memory, 0, region.length) {
         Some(cells) => {
-            // In case you want to do some premature optimization, this shows how to cast a `&'mut [Cell<u8>]` to `&mut [u8]`:
-            // https://github.com/wasmerio/wasmer/blob/0.13.1/lib/wasi/src/syscalls/mod.rs#L79-L81
+            let raw_cells = cells as *const [_] as *const u8;
             let len = region.length as usize;
             let mut result = vec![0u8; len];
```

Comment:
Does Rust allow us to create an a vector of the correct length without zeroing it? There is no point in wrinting all zeros here and then overwriting it in the next line.

Reply:
Seems that's the preferred, and one the fastest ways. See [https://github.com/rust-lang/rust/issues/54628](https://github.com/rust-lang/rust/issues/54628). For what they say, `vec!` has special behaviour when its first argument is zero.

Reply:
It might be the best for initialized memory. But we only need uninitialized memory that just holds any data from previous memory use instead of writing zeros there.

We can do

```rust
// Allocate vector big enough for len elements.
let mut result = Vec::with_capacity(len);

// write into the vector

// Mark the first len elements of the vector as being initialized.
unsafe {
    result.set_len(len);
}
```

See also this example: https://doc.rust-lang.org/std/vec/struct.Vec.html#examples-18.

Then we write to this memory only once instead of twice.

Reply:
Nice.

I was thinking, on a related note, do we want to do de the same for `write_region`?

Reply:
I think we should do both in parallel, in one PR. Feel free to cherry pick the work from here and open a new PR including both.

Reply:
OK, I can do that, and in fact already have the branch created for the benchmarks.

Another option would be for @slave5vw to add these changes, along with the requested ones above, and so it gets the credit directly.

@slave5vw, what do you think?

Reply:
In any case, I've already implemented these, in the [read_region-copy branch](https://github.com/CosmWasm/cosmwasm/tree/read_region-copy), in order to [benchmark it](https://app.circleci.com/pipelines/github/CosmWasm/cosmwasm?branch=read_region-copy).